src/rules/no_unexpected_multiline.rs

```diff
@@ -0,0 +1,322 @@
+  }
+}
+
+struct NoUnexpectedMultilineVisitor {
+  context: Arc<Context>,
+}
+
+impl NoUnexpectedMultilineVisitor {
+  pub fn new(context: Arc<Context>) -> Self {
+    Self { context }
+  }
+
+  fn check_for_break_after(&self, outer: Span, after: Span, msg: &str) {
+    lazy_static! {
+      static ref MSG_MAP: HashMap<&'static str, &'static str> = {
+        let mut map = HashMap::new();
+        map.insert(
+          "call",
+          "Unexpected newline between function and ( of function call",
+        );
+        map.insert(
+          "member",
+          "Unexpected newline between object and [ of property access",
+        );
+        map.insert(
+          "div",
+          "Unexpected newline between numerator and division operator",
+        );
+        map
+      };
+    }
+    let source_map = &self.context.source_map;
+    let before_paren = outer.trim_start(after).unwrap();
+    let temp_span = source_map
+      .span_take_while(before_paren, |c| *c == ')' || c.is_whitespace());
+    let left_paren = before_paren.trim_start(temp_span).unwrap();
+    let line1 = source_map.lookup_char_pos(after.hi()).line;
+    let line2 = source_map.lookup_char_pos(left_paren.lo()).line;
+    if line1 != line2 {
+      self.context.add_diagnostic(
+        left_paren.with_lo(left_paren.lo() + BytePos(1)),
+        "no-unexpected-multiline",
+        MSG_MAP[msg],
+      );
+    }
+  }
+}
+
+impl Visit for NoUnexpectedMultilineVisitor {
+  fn visit_bin_expr(
+    &mut self,
+    bin_expr: &swc_ecma_ast::BinExpr,
+    _parent: &dyn Node,
+  ) {
+    lazy_static! {
+      static ref SLASH_AND_FLAGS: regex::Regex =
+        Regex::new(r"^/[gimsuy]+(?:[\W].*)?$").unwrap();
+    }
+    self.visit_expr(&bin_expr.left, _parent);
+    self.visit_expr(&bin_expr.right, _parent);
+
+    if let swc_ecma_ast::BinaryOp::Div = bin_expr.op {
+      if let swc_ecma_ast::Expr::Bin(inner_bin_expr) = &*bin_expr.left {
+        let temp_span = bin_expr.span.trim_start(bin_expr.left.span()).span();
+        let source_map = &self.context.source_map;
+        let slash_and_flags = source_map
+          .span_to_snippet(
+            temp_span
+              .trim_start(source_map.span_take_while(temp_span, |c| *c != '/'))
+              .unwrap(),
+          )
+          .unwrap();
+        if !matches!(inner_bin_expr.op, swc_ecma_ast::BinaryOp::Div)
+          || !SLASH_AND_FLAGS.is_match(&slash_and_flags)
+        {
+          return;
+        }
+        self.check_for_break_after(
+          bin_expr.span,
+          inner_bin_expr.left.span(),
+          "div",
+        );
+      }
+    }
+  }
+
+  fn visit_call_expr(
+    &mut self,
+    call_expr: &swc_ecma_ast::CallExpr,
+    _parent: &dyn Node,
+  ) {
+    if let swc_ecma_ast::ExprOrSuper::Expr(expr) = &call_expr.callee {
+      self.visit_expr(expr, _parent);
+    }
+    for arg in &call_expr.args {
+      self.visit_expr(&arg.expr, _parent);
+    }
+    // TODO: handle ES2020 optional chaining
```

Comment:
What's exactly missing here? SWC has `Expr::OptChain` which gives [`OptChainExpr`](https://docs.rs/swc_ecma_ast/0.25.0/swc_ecma_ast/struct.OptChainExpr.html), I think that could be used?

Reply:
Specifically I would need the ability to differentiate between what's before and after the `?.` operator. For example, according to ESLint in the following the first is lint free and the second has an unexpected-multiline:
```
foo.bar?.
(a);

foo.bar
(a)?.b;
```
In SWC's `OptChainExpr` in both cases the value of `expr` would be the whole expression.

Reply:
I'm still not entirely grasping the problem... 😒  but it seems the problem lies in SWC after all. @qkniep could you provide example what would work in this case?

CC @kdy1 for opinion

Reply:
I think OptChainExpr should have a field like `question_dot_token: Span`.
Will adding it fix the issue?

Reply:
Adding `question_dot_token: Span would allow to check the line this token is on, so it should be possible to compare lines between statements in the above example. It could work 👍